### DISCUSSION TITLE
[Feature]: Metadata Editor

### DISCUSSION THREAD

**Discussion by Nonobis** [immich-app/immich#1696] [Feature Request]
### Feature detail

Hello,

A metadata editor could be very cool feature, with update of metadata directly on file

For fixing date/time of files or gps coordinate etc ...

### Platform

Web

**Comment by r3nor** [comment]
This would be great!

**Comment by panoti** [comment]
For this issue, I suggest some solutions:
1. Wrap `native` tools like [exiftool](https://exiftool.org/), [exiv2](https://exiv2.org/) by using [childprocess](https://nodejs.org/api/child_process.html) to parse and write metadata.
    - **Pros**: so easy, best performance, always updated, support many file types.
    - **Cons**: heavily depend on perl (only with exiftool), install more binary packages (exiv2)
2. Wrap `native` libs like `exiv2` by using [node-gyp](https://github.com/nodejs/node-gyp). Exiv2 also provide API so that other app can communicate to.
    - **Pros**: best performance, always updated, support many file types.
    - **Cons**: not easy to implement with node-gyp.
3. Transpile `native` libs like `exiv2` to wasm.
    - **Pros**: best performance, always updated, support many file types.
    - **Cons**: pretty difficult to transpile C++ to wasm.
4. Using `pure` javascript lib but they're **outdated** like [piexifjs](https://github.com/hMatoba/piexifjs).
    - **Pros**: so easy
    - **Cons**: outdated, lack of support, performance issues.

**Comment by JVT038** [comment]
Additionally, can you also add a feature for mass metadata editing?

I have thousands of videos and images from WhatsApp and they're all in `IMG-YYYYMMDD-WA0001.JPG` format, but don't have any exif data, so Immich automatically assumes they're taken on the file creation date (which in my case is March 2022). So there are currently a lot of images from 2017, 2019, 2020, etc. which are placed in March 2022, because they're from WhatsApp without exif data.
I'm currently fixing this with the exiftool, which takes the filename and sets the exif data based on the filename, but could this be done in the web ui? I'm sure I'm not the only one uploading WhatsApp images to Immich

**Reply by pablo03v** [reply]
yes, same situation for me. I uploaded like 300 photos but there date is wrong. A possiblity to change the date would certainly be great!

**Reply by nielstron** [reply]
I recently built a specialized tool for this here: https://github.com/nielstron/exif_from_filename
Requires a bit of technical skill to operate though (i.e. connect to server/mount storage locally, run in CLI)

**Comment by r3nor** [comment]
I think that, following on what @JVT038 said, it would be a very good feature to fallback on filename timestamp (if available) when no metadata is found. Most images from phones or cameras contain the timestamp on the filename, so it would be a good fallback before using the modified date.

**Comment by jkaplon** [comment]
Might this also include a "Description" field?

gphotos has it in the metadata panel, and it's proven useful with pics of kid drawings (where there's no chance of success by an image classifier).

Hopefully this is a well-supported EXIF item and will be easy to add...but I don't want to be making scope-creep-assumptions.

**Comment by samip5** [comment]
The exif parsing is actually done with exiftool-vendored library now. :)

**Comment by AD-Wright** [comment]
I'm looking for some way to edit the date/time of already-uploaded photos, and I think for that purpose editing the exif data would be useful.  Was this issue closed by mistake?  I can't find any way to edit the metadata.  
Also, I'm curious - if the date/time was changed, would the photo be moved to the proper place in the folder structure?

**Comment by jrasm91** [comment]
Related to #2466, which adds support for sidecar files, with the option to re-extract exif from them.

**Reply by feerlessleadr** [reply]
So theoretically, I could use exiftool to fix the exifdata of the files, use exiftool to generate the xmp with the correct info next to the already imported picture, then have immich rescan the library and reimport the exif data using the newly generated xmp sidecar file?

**Reply by jrasm91** [reply]
It's still being finalized, but I think the better approach would be to not update the originals directly (this impacts the hash and causes all sorts of issues), but instead edit the exif and save it instead to a sidecar file. Immich will import and apply the new data from the file over the values from the original file.

**Reply by feerlessleadr** [reply]
Thanks, I'll have to figure out a way to do that with exiftool, or if there is a better tool to do that.

**Reply by jrasm91** [reply]
Exiftool can do it. I think there are lots of 3rd party tools for editing and most should support writing to a sidecar. The OP of the PR mentioned digikam

**Reply by feerlessleadr** [reply]
Thanks, when I was reading I guess I figured it only exported to sidecar rather direct writing to a sidecar.

**Reply by jrasm91** [reply]
Yup, it should be able to do both

**Comment by iqapps** [comment]
May I suggest, that when downloading photos, you are given the option to download the original, or a modified with overwritten exif from sidecar data.

**Comment by MalteKiefer** [comment]
Any news here on the status?

**Reply by bo0tzz** [reply]
No news

**Comment by silverl** [comment]
I just completed my first enormous Google Takeout-to-Immich import. So far so good!

Oddly, I have 3 files that for whatever reason came in with a year of 2036. Now they always show at the top of my gallery.

May I ask, what is the correct way to fix a date in the absence of a metadata editor?

Thanks

**Reply by alextran1502** [reply]
You can edit the exif of the file using exif-tool then rerun the metadata extraction job

**Reply by jrasm91** [reply]
No super recommended though as the hash will be out of sync. Better to delete them, correct the exit and re-upload. Or, use a tool that can write sidecar xmp files, which immich can detect.

**Reply by eygraber** [reply]
@jrasm91 what are the implications of the hash being out of sync? I have thousands of images that need to be corrected, and I'm not looking forward to deleting, correcting, and then re-uploading.

**Reply by jrasm91** [reply]
Not a big deal now as it is only used for duplicates but it may be used for integrity checks in the future. Another option is to use sidecar files instead of updating the original directly

**Reply by GuoxiongXie** [reply]
Thanks Alex, when you said "You can edit the exif of the file using exif-tool", do you mean directly editing those files inside UPLOAD_LOCATION/library folder? Or do you mean editing some files outside of immich file system, and then upload those edited files to immich? Thanks for your clarifications!

**Comment by bcsteeve** [comment]
I don't know if it is new or if I never noticed, but Google Photos has the ability to edit date/time and location!  If that's been like that for awhile, I don't know how I missed it, because I went through a lot of effort looking for an exiftool GUI to avoid learning the command line :)

**Comment by HarHarLinks** [comment]
I would like to be able to
- add missing metadata such as location
- edit metadata such as the timestamp e.g. in cases where the fallback timestamp is wrong and the original timestamp is (roughly) known
- add tags of persons who could not be automatically identified for any reason, but I am unsure that fits inside this suggestion

I'm sure that could be done in a way that only optionally touches the original files.

**Comment by lordraiden** [comment]
a bulk mode would be nice, to edit many files at once

**Comment by PterX** [comment]
This would be great!

**Comment by alien75** [comment]
Agree: that would be a very useful addition!

**Comment by imthefrizzlefry** [comment]
It would be nice to be able to write this data to the image files directly in case one day my decendents/family want to import my photos including updates I painstakingly made to thousands of pictures/videos, but at least being able to update any values stored in the immich database would make that possible as a future feature.

As an end user adding photos, 
I want the ability to modify metadata, 
So I can manage my content collection.

Use Cases:
A number of images imported from a previous photo management system had an incorrect timestamp; I want to update the timestamp via mobile/web.
An image I scanned has the incorrect timestamp, I want to update the timestamp via web/mobile.
An image I scanned has no location data; I want to update the location via web/mobile.
An image I took with my phone has little meaning if I forget the purpose of the photo; I want to add a comment to the image metadata so I never forget.
A Video I digitized from from reel/vhs has the incorrect timestamp; I want to update the timestamp via web/mobile.

**Comment by TiloHeidasch** [comment]
To me (and appearantly not only me) this is the most important feature missing from Immich right now!

I was about to create my own feature request before stumbling on this one, so if it is helpful, here's the user story from my feature request:

> As an authorized user,
> I want to change metadata (Date when picture was taken / Place on Map where picture was taken) of my indexed images,
> so I can correct wrong or missing metadata after having them indexed by Immich.

I have several Photos, that have been digitized after the fact using a scanner. These Images date back to 1920. However, the date where they appear in Immich is of course their digitization date, which was more in the realm of 2015. I want to change the "Date where Immich sorts them to in the timeline" to the "correct" date manually.

I have several Photos that do not have / have a false location. They don't appear on the map view and some appear in their falsely marked locations e.g. 0,0 . I want to change the "Location where the picture is shown on the map" by clicking on a respective point on a map.

I have several Photos that require manual intervention, therefor I want the edit operations to be able to be done in bulk for as many Items I select. When selecting more than 1 Photo and changing any of the beforementioned metadata, I want all the photos to get the changed attribute.

**Reply by HarHarLinks** [reply]
I too have the 0,0 problem and similar problems where my phone's camera was not able to retrieve the correct location that I'd like to fix. The way to do that would be to select one or more photos to edit and then select the location on the map. The map should have a search box so I can easier navigate the world to some sight from my summer holiday trip I would have a hard time finding on the map. It should also allow me to select existing photos' locations to copy their location data.

**Comment by YFrendo** [comment]
I start a Draft, don't expect to much for now but if you want to follow my work : #5066

**Reply by AngelaDMerkel** [reply]
This is awesome. I think a toggle allowing the user to choose to write to a new sidecar or overwrite existing, or write to the image would both be very welcome

**Reply by YFrendo** [reply]
I choosed the sidecar path. 
Write directly to the image is more tricky and can cause a lot of new issue (change the hash...).
For now the server part is ready for change date and position for singular asset or in bulk.

For now it overide the preexisting sidecar (but create a save of the previous one), I assume that if a user changes the metadata he also wants to modify his sidecar. I don't see any case where that wouldn't be the case.

I need to make the frontend part now x)

**Comment by jrasm91** [comment]
You can now update gps and date/time on the web and mobile app.

**Reply by HarHarLinks** [reply]
You can, but it is incomplete yet.
E.g. adding og people is missing, and the location editing map UI can be improved a lot.

**Reply by imthefrizzlefry** [reply]
I'll take a partial/imperfect implementation over nothing.  Progress is good!  Thank you for the hard work!  We can incrementally improve it over time!

**Reply by HarHarLinks** [reply]
I definitely agree, but I wouldn't have closed this discussion yet.

**Comment by Sums123** [comment]
I made a post in relation to the timzeone edit functionality here
https://github.com/immich-app/immich/discussions/6268
